### PR TITLE
Add constructvariable! test

### PR DESCRIPTION
**Pull request by blegat**

### CODE REVIEW

File: test/variable.jl [hunk 1]
```diff
@@ -14,6 +14,21 @@ using JuMP
 import JuMP.repl
 using Base.Test
 
+mutable struct MyVariable
+    haslb::Bool
+    lowerbound::Number
+    hasub::Bool
+    upperbound::Number
+    hasfix::Bool
```
Comment (mlubin) [inline test/variable.jl]: The documentation in `macros.jl` of `constructvariable!` is not in sync with this (it doesn't mention `haslb`, `hasub`, `hasfix`, `fixedvalue`, or `hasstart`). This whole interface with positional arguments is very confusing. One possible improvement is to have the macro pass a `VariableInfo` struct to `constructvariable!` where `VariableInfo` is defined as
```julia
struct VariableInfo
    haslb::Bool
￼    lowerbound::Number
￼    hasub::Bool
￼    upperbound::Number
￼    hasfix::Bool
    fixedvalue::Number
￼    binary::Bool
￼    integer::Bool
￼    name::String
￼    hasstart::Bool
￼    start::Number
end
```
(maybe parameterized on the number type for performance?)
This way the keyword arguments are separated from "special" macro syntax.

Comment (blegat) [reply test/variable.jl]: Done :)

File: src/macros.jl [hunk 2]
```diff
@@ -873,34 +873,45 @@ esc_nonconstant(x::Number) = x
...
+end
+
 # Returns the type of what `constructvariable!` would return with these starting positional arguments.
 variabletype(m::Model) = Variable
 # Returns a new variable belonging to the model `m`. Additional positional arguments can be used to dispatch the call to a different method.
 # The return type should only depends on the positional arguments for `variabletype` to make sense.
-function constructvariable!(m::Model, _error::Function, haslb::Bool, lowerbound::Number, hasub::Bool, upperbound::Number,
-                            hasfix::Bool, fixedvalue::Number, binary::Bool, integer::Bool, name::String,
-                            hasstart::Bool, start::Number; extra_kwargs...)
+function constructvariable!(m::Model, _error::Function, info::VariableInfo, name::String; extra_kwargs...)
```
Comment (mlubin) [inline src/macros.jl]: This is much better. What's the motivation for leaving `name` out of `VariableInfo`?

File: src/macros.jl [hunk 3]
```diff
@@ -939,17 +951,17 @@ end
...
 # * The keyword arguments start, basename, lowerbound, upperbound, binary, and integer category may not be passed as is to
 #   `constructvariable!` since they may be altered by the parsing of `expr` and we may need to pass it pointwise if it is a container since
-#   `constructvariable!` is called separately for each variable of the container. Moreover it will be passed as positional argument to `constructvariable!`.
+#   `constructvariable!` is called separately for each variable of the container. Moreover they will be passed inside the VariableInfo object.
 # * A custom error function is passed as positional argument to print the full @variable call before the error message.
 #
-# Examples (... is the custom error function):
-# * `@variable(m, x >= 0)` is equivalent to `x = constructvariable!(m, msg -> error("In @variable(m, x >= 0): ", msg), true, 0, false, NaN, false, NaN, false, false, "x", false, NaN)
+# Examples:
+# * `@variable(m, x >= 0)` is equivalent to `x = constructvariable!(m, msg -> error("In @variable(m, x >= 0): ", msg), VariableInfo(true, 0, false, NaN, false, NaN, false, false, false, NaN), "x")
```
Comment (mlubin) [inline src/macros.jl]: "x" should be passed to `VariableInfo`, here and below.

**Pull request merged**